### PR TITLE
Create python binding

### PR DESCRIPTION
**Pull request by yonidavidson**
I created a lib that can be read from python.
This will be used by a python container to be read using ctypes.
With that we can use it from Python.

### CODE REVIEW

**Comment by yonidavidson** [comment]
fix #4

**Comment by yonidavidson** [comment]
@tebeka Hi, can you have a look please? I am not sure but I can't link python-bindings with the lib_arrow lib. Can you see if I am doing anything very wrong?
Thanks

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,5 @@ carrow
 *.a
 .vscode
 .idea
+lib/artifacts/*
```
Comment (tebeka) [inline .gitignore]: Why do we need these two directories?

File: lib/carrow_lib.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
+package main
+
+import (
+	"C"
+	"fmt"
+	"unsafe"
+	"github.com/353solutions/carrow"
```
Comment (tebeka) [inline lib/carrow_lib.go]: Should be in it's own import group

File: python-bindings/setup.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+ext_modules = cythonize("example.pyx")
+
+for ext in ext_modules:
+    # The Numpy C headers are currently required
+    ext.include_dirs.append(np.get_include())
+    ext.include_dirs.append(pa.get_include())
+    ext.libraries.extend(pa.get_libraries())
+    ext.library_dirs.extend(pa.get_library_dirs())
+    ext.include_dirs.append("/home/carrow/lib/artifacts/")
```
Comment (tebeka) [inline python-bindings/setup.py]: Don't hard code paths. You can use
```
from os.path import dirname, abspath
here = dirname(abspath(__file__))

...
ext.include_dirs.append(f'{here}/lib/atrifacts')
```

File: python-bindings/example.pyx [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# distutils: language=c++
+
+from pyarrow.lib cimport *
+
+
+def get_array_length(obj):
```
Comment (tebeka) [inline python-bindings/example.pyx]: We usually don't have `get_` prefix in Python. Also the example itself should be pure Python and using the `carrow` Python extension module.

**Comment by tebeka** [comment]
The way I see the code structured is:
- Main code is a Go module that exposes Arrow API
- In a different folder (`py`?) we have a Python extension module that uses the Go module to interact with Arrow. It should compile the Go to DLL and use Cython to connect to it.

File: Dockerfile [hunk 5]
```diff
@@ -24,6 +24,15 @@ RUN apt install -y -V libgandiva-glib-dev
 RUN apt install -y -V libparquet-dev 
 RUN apt install -y -V libparquet-glib-dev
 
+# Python bindings
+RUN apt install -y -V curl
```
Comment (tebeka) [inline Dockerfile]: I perfer to install miniconda for this.

```
RUN curl -LO https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
RUN bash -b -p /conda && /conda/bin/conda update -n base conda
ENV PATH=${PATH}:/conda/bin
```

File: Dockerfile [hunk 6]
```diff
@@ -24,6 +24,15 @@ RUN apt install -y -V libgandiva-glib-dev
...
+
+RUN cd /tmp && \
```
Comment (tebeka) [inline Dockerfile]: Use `conda` (see above)

File: .circleci/config.yml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+version: 2
```
Comment (tebeka) [inline .circleci/config.yml]: Why is this showing in the diff?

File: python-bindings/Makefile [hunk 8]
```diff
@@ -0,0 +1,12 @@
+.PHONY:
```
Comment (tebeka) [inline python-bindings/Makefile]: I'd like to have a Python project structure. Something like

```
+ py
  |
  +- carrow (all files go here)
  +- setup.py (project definition)
```

File: python-bindings/Makefile [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+all:
+	go build -o libcarrow_bindings.so -buildmode=c-shared carrow_bindings.go && \
+	pip install -r requirements.txt && \
+	python3.6 setup.py build_ext --inplace
```
Comment (tebeka) [inline python-bindings/Makefile]: Just `python`. System setup will decide which Python (3.7?)

File: python-bindings/bindings.pyx [hunk 10]
```diff
@@ -0,0 +1,15 @@
+# distutils: language=c++
+
+from pyarrow.lib cimport *
```
Comment (tebeka) [inline python-bindings/bindings.pyx]: Don't do `import *`, import just what you need.

File: python-bindings/carrow_bindings.go [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+	if err != nil {
+		fmt.Printf("table creation error: %s", err)
+		return
+	}
+
+	fmt.Printf("num cols: %d\n", table.NumCols())
+	fmt.Printf("num rows: %d\n", table.NumRows())
+
+	//return int(table.Pointer())
+	return
```
Comment (tebeka) [inline python-bindings/carrow_bindings.go]: Should return the pointer?

File: python-bindings/Makefile [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+all:
+	go build -o libcarrow_bindings.so -buildmode=c-shared carrow_bindings.go && \
+#	pip install -r requirements.txt && \
+	python setup.py build_ext --inplace
+
+clean:
+	rm -rf *.so *.cpp *.c *.h build
+
+test:
+	ldconfig && python runner.py
```
Comment (tebeka) [inline python-bindings/Makefile]: IMO the `ldconfig` should be in the `Dockerfile`

File: python-bindings/Makefile (hunk 8 above)
Comment (yonidavidson) [reply python-bindings/Makefile]: What do you want in the root folder, only setup.py ?
what about other files like the runner or pyx pxd

File: python-bindings/Makefile (hunk 12 above)
Comment (yonidavidson) [reply python-bindings/Makefile]: is 	LD_LIBRARY_PATH=$${PWD} python runner.py OK?

**Comment by yonidavidson** [comment]
Closed, preferred to open a new #22 after changes were too big.

**Pull request closed without merging**